python/lsst/afw/table/catalogMatches.py

```diff
@@ -254,3 +256,112 @@ def copyAliasMapWithPrefix(inSchema, outSchema, prefix=""):
         outSchema.getAliasMap().set(prefix + k, prefix + v)
 
     return outSchema
+
+
+def reindexCatalog(catalog, indices, deep=True):
+    """Apply a numpy index array to an afw Catalog
+
+    Parameters
+    ----------
+    catalog : subclass of `lsst.afw.table.BaseCatalog`
+        Catalog to reindex.
+    indices : `numpy.ndarray` of `int`
+        Index array.
+
+    Returns
+    -------
+    new : subclass of `lsst.afw.table.BaseCatalog`
+        Reindexed catalog. Records are shallow copies of those in ``catalog``.
+    """
+    new = type(catalog)(catalog.schema)
+    new.reserve(len(indices))
```

Comment:
Probably don't want this `reserve` line, because you're making shallow copies and hence don't need to allocate new memory (there will be memory allocated for the pointers, but I don't think you need to worry about allocating that up-front; `std::vector` will be smart enough to do it plenty efficiently without any help).

Reply:
I think the default is to make deep copies

Reply:
Yeah, I added the `deep` argument to `new.extend` because otherwise it returned non-contiguous catalog that I couldn't do anything with. But left it as a parameter in case it is useful to a user for speed.

Reply:
Are you constructing the catalogs and then accessing them through a numpy interface?

Reply:
Ah, I didn't even notice that you were actually providing the option of doing deep copies below (I just saw "shallow" in the return docs above).

If you want to optionally do deep copies, do this:
```
new = SourceCatalog(catalog.table.clone() if deep else catalog.table)
records = [catalog[int(ii)] for ii in indices]
new.extend(records, deep=deep)
```

Reply:
(and then also document the `deep` parameter)

Reply:
@natelust yes. This was originally intended to be used by the `matchCatalogsExact` function until I learned that `pipe_analysis` requires a `Match` object. But I still think it's useful to be able to go `afwCatalog[indices]` and get back a table that is matched in some way.

Reply:
Jim, does that need to be a list comp or could a generator exp work fine?

Reply:
Probably _better_ for it to be a list or a tuple, because it'll be iterated over in pybind11, so considerations like being able to use [`PySequence_Fast`](https://docs.python.org/3/c-api/sequence.html#c.PySequence_Fast) in the C API are probably more relevant than the usual pure-Python wisdom of preferring lazy iterators.